drivers/mbox/mbox_ti_omap.c

```diff
@@ -0,0 +1,196 @@
+		if (!data->channel_enable[i_channel])
+			continue;
+
+		if ((flags & MAILBOX_IRQ_NEWMSG(i_channel))) {
+			data->received_data = regs->message[i_channel];
+			struct mbox_msg msg = {(const void *)&data->received_data,
+					       MAILBOX_MBOX_SIZE};
+
+			if (data->cb[i_channel]) {
+				data->cb[i_channel](dev, i_channel, data->user_data[i_channel],
+						    &msg);
+				irq_regs->status_clear = MAILBOX_IRQ_NEWMSG(i_channel);
+			}
+
+		}
+	}
+}
+
+static int omap_mailbox_send(const struct device *dev, uint32_t channel, const struct mbox_msg *msg)
+{
+	uint32_t __aligned(4) data32 = 0;
+	struct omap_mailbox_data *data = DEV_DATA(dev);
+	volatile struct omap_mailbox_regs *regs = DEV_BASE(data);
+
+	if (channel >= MAILBOX_MAX_CHANNELS)
+		return -EINVAL;
+
+	if (regs->fifo_status[channel])
+		return -EBUSY;
+
+	if (!msg) {
+		regs->message[channel] = 0;
+		return 0;
+	}
+
+	if (msg->size > MAILBOX_MBOX_SIZE)
+		return -EMSGSIZE;
+
+	memcpy(&data32, msg->data, msg->size);
+	regs->message[channel] = data32;
+
+	return 0;
+}
+
+static int omap_mailbox_register_callback(const struct device *dev, uint32_t channel,
+					  mbox_callback_t cb, void *user_data)
+{
+	struct omap_mailbox_data *data = DEV_DATA(dev);
+
+	if (channel >= MAILBOX_MAX_CHANNELS)
+		return -EINVAL;
+
+	data->cb[channel] = cb;
+	data->user_data[channel] = user_data;
+
+	return 0;
+}
+
+static int omap_mailbox_mtu_get(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	return MAILBOX_MBOX_SIZE;
+}
+
+static uint32_t omap_mailbox_max_channels_get(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	return MAILBOX_MAX_CHANNELS;
+}
+
+static int omap_mailbox_set_enabled(const struct device *dev, uint32_t channel, bool enable)
+{
+	struct omap_mailbox_data *data = DEV_DATA(dev);
+	const struct omap_mailbox_config *cfg = DEV_CFG(dev);
+	volatile struct omap_mailbox_regs *regs = DEV_BASE(data);
+	uint32_t irqstatus;
+
+	if (channel >= MAILBOX_MAX_CHANNELS)
+		return -EINVAL;
+
+	if (enable && data->channel_enable[channel])
+		return -EALREADY;
+
+	irqstatus = regs->irq_regs[cfg->usr_id].enable_set;
+	irqstatus |= MAILBOX_IRQ_NEWMSG(channel);
+	regs->irq_regs[cfg->usr_id].enable_set = irqstatus;
+	data->channel_enable[channel] = enable;
+
+	return 0;
+}
+
+static const struct mbox_driver_api omap_mailbox_driver_api = {
+	.send = omap_mailbox_send,
+	.register_callback = omap_mailbox_register_callback,
+	.mtu_get = omap_mailbox_mtu_get,
+	.max_channels_get = omap_mailbox_max_channels_get,
+	.set_enabled = omap_mailbox_set_enabled,
+};
+
+#define MAILBOX_INSTANCE_DEFINE(idx)	\
+	static struct omap_mailbox_data omap_mailbox_##idx##_data;	\
+	const static struct omap_mailbox_config omap_mailbox_##idx##_config = {	\
+		.usr_id =  DT_INST_PROP(idx, usr_id),	\
+	};	\
+	static int omap_mailbox_##idx##_init(const struct device *dev)	\
+	{	\
+		struct omap_mailbox_data *data = dev->data;	\
+		mem_addr_t addr;	\
+		device_map((mm_reg_t *)&addr, DT_INST_REG_ADDR(idx),	\
```

Comment:
I think the `DEVICE_MMIO_*` macros are preferred for this

Reply:
Thanks for the review @gramsay0 , will fix this and other comments.

Reply:
Done, added to TI K3 MAINTAINERS entry